File: scripts/characters/publish-update-packs.mjs
import { parseArgs, promisify } from 'node:util'
import { execFile } from 'node:child_process'
import { mkdir, mkdtemp, readFile, rm, stat, writeFile } from 'node:fs/promises'
import { dirname, join, resolve } from 'node:path'
import { tmpdir } from 'node:os'
import { pathToFileURL } from 'node:url'
import { withPackTools } from './pack-tools.mjs'
import { archiveHash } from './add-update-source.mjs'

/** Exact-file allowlist. A directory is never an upload operation. No credentials in the plan. */
export async function preparePublication(plan) {
  if (!plan || Object.keys(plan).some(k => !['repoId', 'public', 'packs', 'documents'].includes(k)) || plan.public !== true || !Array.isArray(plan.packs) || plan.packs.length < 1 || plan.packs.length > 32 || !Array.isArray(plan.documents) || plan.documents.length > 8) throw Error('Invalid public publication plan')
  const stage = await mkdtemp(join(tmpdir(), 'petchar-publication-'))
  try { return await withPackTools(async tools => {
    const packs = [], ids = new Set(), paths = new Set()
    for (const [index, item] of plan.packs.entries()) {
      if (!item || Object.keys(item).some(k => !['pack', 'packId', 'artifactPath', 'manifestPath', 'minAppVersion', 'notes'].includes(k))) throw Error('Invalid pack allowlist')
      const root = join(stage, String(index)); await mkdir(root)
      const pack = await tools.extractCharacterPack(resolve(item.pack), root), m = pack.manifest
      if (m.id !== item.packId || ids.has(m.id) || !m.update || m.update.repoId !== plan.repoId || m.update.manifestPath !== item.manifestPath || paths.has(item.artifactPath) || paths.has(item.manifestPath)) throw Error('Pack/source/allowlist mismatch')
      const sha256 = await archiveHash(join(root, 'archive.zip'))
      const feed = tools.parseUpdateFeed(Buffer.from(JSON.stringify({ schemaVersion: 1, packId: m.id, version: m.version, minAppVersion: item.minAppVersion, runtime: m.runtime, artifact: { path: item.artifactPath, revision: '0'.repeat(40), bytes: (await stat(join(root, 'archive.zip'))).size, sha256 }, notes: item.notes })))
      // This template is private and cannot be published until the actual artifact commit is known.
      delete feed.artifact.revision
      if (await archiveHash(item.pack) !== sha256) throw Error('Source archive changed during validation')
      ids.add(m.id); paths.add(item.artifactPath); paths.add(item.manifestPath)
      packs.push({ file: resolve(item.pack), path: item.artifactPath, manifestPath: item.manifestPath, sha256, bytes: feed.artifact.bytes, feed })
    }
    const documents = []
    for (const item of plan.documents) {
      if (!item || Object.keys(item).length !== 2 || typeof item.file !== 'string' || !['README.md', 'SHA256SUMS.txt', 'PROVENANCE.md', 'LICENSE.txt'].includes(item.path) || paths.has(item.path)) throw Error('Document not allowlisted')
      if ((await stat(item.file)).size > 128 * 1024) throw Error('Document too large')
      paths.add(item.path); documents.push({ file: resolve(item.file), path: item.path, sha256: await archiveHash(item.file) })
    }
    return { repoId: plan.repoId, public: true, packs, documents }
  }) } finally { await rm(stage, { recursive: true, force: true }) }
}
if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  const { values } = parseArgs({ options: { plan: { type: 'string' }, record: { type: 'string' }, publish: { type: 'boolean', default: false }, 'create-repo': { type: 'boolean', default: false }, python: { type: 'string', default: process.platform === 'win32' ? 'python' : 'python3' } } })
  if (!values.plan || !values.record) throw Error('Require --plan <exact-file JSON> --record <private resumable record>; default is dry-run')
  if ((await stat(values.plan)).size > 128 * 1024) throw Error('Plan too large')
  const plan = await preparePublication(JSON.parse(await readFile(values.plan, 'utf8')))
  console.log(JSON.stringify({ mode: values.publish ? 'publish' : 'dry-run', destination: `https://huggingface.co/datasets/${plan.repoId}`, visibility: 'public, ungated', files: [...plan.packs.flatMap(p => [p.path, p.manifestPath]), ...plan.documents.map(d => d.path)], hashes: plan.packs.map(p => ({ path: p.path, sha256: p.sha256, bytes: p.bytes })) }, null, 2))
  if (values.publish) {
    const staging = await mkdtemp(join(tmpdir(), 'petchar-publish-plan-'))
    try {
      const validated = join(staging, 'plan.json'); await writeFile(validated, JSON.stringify(plan), { mode: 0o600 })
      await mkdir(dirname(resolve(values.record)), { recursive: true })
      const result = await promisify(execFile)(values.python, [join(import.meta.dirname, 'publish-update-packs.py'), '--plan', validated, '--record', resolve(values.record), '--publish', ...(values['create-repo'] ? ['--create-repo'] : [])], { timeout: 30 * 60_000, maxBuffer: 1024 * 1024, env: { ...process.env, HF_HUB_DISABLE_TELEMETRY: '1', HF_HUB_DISABLE_PROGRESS_BARS: '1' } })
      console.log(result.stdout)
    } catch { throw Error('Publication did not complete. Check the private record; no feed is published before artifact verification. HF write access and the developer huggingface_hub package are required.') }
    finally { await rm(staging, { recursive: true, force: true }) }
  }
}

File: scripts/characters/publish-update-packs.py
"""Official HF API publisher, invoked only with a fully validated exact-file plan.

Developer dependency: huggingface_hub. Never shipped as an app dependency.
No token arguments, repository-wide uploads, deletion, or visibility changes.
"""
import argparse
import hashlib
import json
import os
from pathlib import Path
import tempfile
import urllib.parse
import urllib.request

HOSTS = {"huggingface.co", "cdn-lfs.huggingface.co", "cdn-lfs.hf.co", "cdn-lfs-us-1.hf.co", "cdn-lfs-eu-1.hf.co", "cas-bridge.xethub.hf.co", "us.aws.cdn.hf.co", "us.gcp.cdn.hf.co", "cas-server.xethub.hf.co", "cas-server.xethub-eu.hf.co", "transfer.xethub.hf.co", "transfer.xethub-eu.hf.co"}


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, *args):
        return None


def anonymous(repo, revision, path, limit, collect=False):
    url = f"https://huggingface.co/datasets/{repo}/resolve/{revision}/{path}"
    opener = urllib.request.build_opener(urllib.request.ProxyHandler({}), NoRedirect())
    for _ in range(6):
        u = urllib.parse.urlsplit(url)
        if u.scheme != "https" or u.hostname not in HOSTS or u.port not in (None, 443) or u.username or u.password or u.fragment:
            raise RuntimeError("DISALLOWED_REDIRECT")
        try:
            response = opener.open(urllib.request.Request(url, headers={"Accept-Encoding": "identity"}), timeout=30)
        except urllib.error.HTTPError as error:
            if error.code in (301, 302, 303, 307, 308):
                url = urllib.parse.urljoin(url, error.headers["Location"])
                error.close()
                continue
            raise RuntimeError(f"ANONYMOUS_HTTP_{error.code}") from None
        with response:
            digest, size, chunks = hashlib.sha256(), 0, []
            while chunk := response.read(1024 * 1024):
                size += len(chunk)
                if size > limit:
                    raise RuntimeError("SIZE_MISMATCH")
                digest.update(chunk)
                if collect:
                    chunks.append(chunk)
            return size, digest.hexdigest(), b"".join(chunks)
    raise RuntimeError("REDIRECT_LIMIT")


def publish(plan, record_path, create_repo):
    from huggingface_hub import HfApi, CommitOperationAdd
    from huggingface_hub.errors import RepositoryNotFoundError
    api = HfApi()
    api.whoami()  # Uses existing local authentication; never prints credentials.
    repo = plan["repoId"]
    try:
        info = api.repo_info(repo, repo_type="dataset")
    except RepositoryNotFoundError:
        if not create_repo:
            raise RuntimeError("BLOCKED_DESTINATION") from None
        api.create_repo(repo, repo_type="dataset", private=False)
        info = api.repo_info(repo, repo_type="dataset")
    if info.private or info.gated:
        raise RuntimeError("REQUIRES_PUBLIC_UNGATED_REPOSITORY")
    record = json.loads(record_path.read_text()) if record_path.exists() else {"repoId": repo, "packs": {}}
    if record["repoId"] != repo:
        raise RuntimeError("RECORD_DESTINATION_CONFLICT")

    def save():
        temporary = record_path.with_suffix(".tmp")
        temporary.write_text(json.dumps(record, indent=2) + "\n")
        os.chmod(temporary, 0o600)
        temporary.replace(record_path)

    paths = set(api.list_repo_files(repo, repo_type="dataset", revision=info.sha))
    operations = []
    for item in plan["packs"]:
        digest = hashlib.file_digest(open(item["file"], "rb"), "sha256").hexdigest()
        if digest != item["sha256"]:
            raise RuntimeError("LOCAL_BYTES_CHANGED")
        previous = record["packs"].get(item["path"])
        if previous and previous["sha256"] != digest:
            raise RuntimeError("IMMUTABLE_VERSION_CONFLICT")
        if item["path"] in paths:
            size, sha, _ = anonymous(repo, info.sha, item["path"], item["bytes"])
            if (size, sha) != (item["bytes"], digest):
                raise RuntimeError("IMMUTABLE_VERSION_CONFLICT")
        else:
            operations.append(CommitOperationAdd(path_in_repo=item["path"], path_or_fileobj=item["file"]))
    for doc in plan["documents"]:
        if hashlib.file_digest(open(doc["file"], "rb"), "sha256").hexdigest() != doc["sha256"]:
            raise RuntimeError("LOCAL_DOCUMENT_CHANGED")
        operations.append(CommitOperationAdd(path_in_repo=doc["path"], path_or_fileobj=doc["file"]))
    artifact_commit = api.create_commit(repo, repo_type="dataset", operations=operations, commit_message="Add validated character pack artifacts", parent_commit=info.sha).oid if operations else info.sha
    record["artifactCommit"] = artifact_commit
    save()
    for item in plan["packs"]:
        size, sha, _ = anonymous(repo, artifact_commit, item["path"], item["bytes"])
        if (size, sha) != (item["bytes"], item["sha256"]):
            raise RuntimeError("ARTIFACT_READBACK_FAILED")
        record["packs"][item["path"]] = {"sha256": sha, "bytes": size, "artifactCommit": artifact_commit, "anonymous": "PASS"}
        save()
    # Recheck remote feeds at a fresh HEAD. Concurrent writes use an expected parent.
    head = api.repo_info(repo, repo_type="dataset").sha
    current_paths = set(api.list_repo_files(repo, repo_type="dataset", revision=head))
    feeds, generated = [], {}
    for item in plan["packs"]:
        feed = item["feed"]
        feed["artifact"]["revision"] = artifact_commit
        if item["manifestPath"] in current_paths:
            _, _, raw = anonymous(repo, head, item["manifestPath"], 65536, True)
            previous = json.loads(raw)
            if previous["packId"] != feed["packId"]:
                raise RuntimeError("REMOTE_FEED_ID_CONFLICT")
            before = tuple(map(int, previous["version"].split(".")))
            after = tuple(map(int, feed["version"].split(".")))
            if before > after or before == after and (previous["artifact"]["sha256"] != item["sha256"] or previous["artifact"]["bytes"] != item["bytes"]):
                raise RuntimeError("REMOTE_FEED_VERSION_CONFLICT")
            if before == after:
                # Preserve the original immutable commit on a resumable equal-version publication.
                feed = previous
        raw = (json.dumps(feed, ensure_ascii=False, indent=2) + "\n").encode()
        generated[item["manifestPath"]] = raw
        feeds.append(CommitOperationAdd(path_in_repo=item["manifestPath"], path_or_fileobj=raw))
    feed_commit = api.create_commit(repo, repo_type="dataset", operations=feeds, commit_message="Publish verified character pack feeds", parent_commit=head).oid
    record["feedCommit"] = feed_commit
    save()
    for path, expected in generated.items():
        _, _, actual = anonymous(repo, feed_commit, path, 65536, True)
        if json.loads(actual) != json.loads(expected):
            raise RuntimeError("FEED_READBACK_FAILED")
        target = record_path.parent / "published-feeds" / path
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(actual)
    record["feedAnonymousReadback"] = "PASS"
    save()
    return {"repoId": repo, "artifactCommit": artifact_commit, "feedCommit": feed_commit, "anonymousReadback": "PASS"}


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--plan", required=True)
    parser.add_argument("--record", required=True)
    parser.add_argument("--publish", action="store_true")
    parser.add_argument("--create-repo", action="store_true")
    args = parser.parse_args()
    if not args.publish:
        raise SystemExit("Dry-run: publication requires the validated Node helper and --publish")
    try:
        print(json.dumps(publish(json.loads(Path(args.plan).read_text()), Path(args.record), args.create_repo)))
    except Exception as error:
        # Third-party exceptions can contain signed URLs. Never echo their message or traceback.
        print(json.dumps({"status": "BLOCKED_PUBLICATION", "errorType": type(error).__name__}))
        raise SystemExit(1)
